Replace formatio with node's util.inspect

mroderick opened the pull request:
This PR fixes #181, it replaces `@sinonjs/formatio` with Node's `util.inspect`

#### Purpose

See if we can drop `@sinonjs/formatio` so it can be deprecated and we don't have to do as much maintenance.

#### Solution

Use Node's `util.inspect` as the object formatter


#### How to verify

1. Observe tests passing

mroderick commented:
@mantoni as you can see, there are two failing tests left.

```
 1) assert.equals
       should fail fail for multi-line strings:

      AssertionError [ERR_ASSERTION]: Message was [assert.equals] "'Yo!\\nMultiline'" expected to be equal to "'Yo!\\nHey'", expected [assert.equals] 'Yo!\nMultiline' expected to be equal to 'Yo!\nHey'
      + expected - actual

      -[assert.equals] "'Yo!\\nMultiline'" expected to be equal to "'Yo!\\nHey'"
      +[assert.equals] 'Yo!\nMultiline' expected to be equal to 'Yo!\nHey'

      at Context.test (lib/test-helper/index.js:126:20)
      at processImmediate (internal/timers.js:439:21)

  2) assert.equals
       should fail fail for multi-line strings with more than one newline:

      AssertionError [ERR_ASSERTION]: Message was [assert.equals] 'Yo!\\nMulti-\\nline' expected to be equal to 'Yo!\\nHey', expected [assert.equals] 'Yo!\nMulti-\nline' expected to be equal to 'Yo!\nHey'
      + expected - actual

      -[assert.equals] 'Yo!\\nMulti-\\nline' expected to be equal to 'Yo!\\nHey'
      +[assert.equals] 'Yo!\nMulti-\nline' expected to be equal to 'Yo!\nHey'

      at Context.test (lib/test-helper/index.js:126:20)
      at processImmediate (internal/timers.js:439:21)
```

Do you know how to fix these?

mantoni commented:
I would say the new formatting of these messages is better than it used to be. We can copy the actual output and use it as the expectation 😊

mroderick commented:
> I would say the new formatting of these messages is better than it used to be. We can copy the actual output and use it as the expectation 😊

I tried that, but something, somewhere removes some of the `\` characters

mantoni commented:
Sounds like there is more code to delete ✌️

mroderick commented:
Here's a challenge: the tests are failing in Node 8 and Node 10, because the `util.inspect` in those is not as capable as the one I used in Node 12, and gives different output.

Some ideas for how to solve this:

1. Make a different set of expectations for each major Node LTS version. It'll be time consuming up front, but will be quite stable going forwards. We'll need to make a set of expectations for https://github.com/browserify/node-util as well, as that has the `util.inspect` from Node 8.
1. Import https://github.com/Automattic/util-inspect, which has not seen an update since 2016 😢 
1. Import https://github.com/browserify/node-util, which has `util.inspect` from Node 8 LTS, and stick with the output from that until `inspect` in `node-util` get's updated.
1. Stop running tests in Node 8, Node 10. Then we might as well drop browser support also. Not a fan of this option
1. `<your suggestions here>`

Ideas?

mantoni commented:
Hmm, part of the idea to use `util.inspect` was to use the „native“ implementation from whatever Node version. I‘d love to have Sinon serialize objects in messages the same way as `console.log`. Therefore option 1 sounds most compelling. Option 3 would kind of achieve the same with less work, but we’d have an „older“ implementation with potentially slightly different output.

mroderick commented:
> Hmm, part of the idea to use `util.inspect` was to use the „native“ implementation from whatever Node version. I‘d love to have Sinon serialize objects in messages the same way as `console.log`. Therefore option 1 sounds most compelling. Option 3 would kind of achieve the same with less work, but we’d have an „older“ implementation with potentially slightly different output.

I suppose we could stub out `util.inspect` and verify that it gets called and that the return value is used?

```js
var inspectStub = sinon.stub(util, "inspect").value("some-uuid");

assert(inspectStub.calledOnce);
assert.equal(
  error.message,
  "[refute.className] Expected object's className not to include " + inspectStub.returnValues[0]
);

mroderick commented:
> I suppose we could stub out `util.inspect` and verify that it gets called and that the return value is used?

That's not going to work very well, considering how the API is assembled.

Perhaps a better approach would be to refactor the tests a bit. In the example of `isMap`,  if we look at what the file actually does:

```js
"use strict";

var actualMessageValues = require("../actual-message-values");

module.exports = function(referee) {
    referee.add("isMap", {
        assert: function(actual) {
            return actual instanceof Map;
        },
        assertMessage: "${customMessage}Expected ${actual} to be a Map",
        refuteMessage: "${customMessage}Expected ${actual} not to be a Map",
        expectation: "toBeMap",
        values: actualMessageValues
    });
};
```

Testing this should be fairly straightforward, and doesn't need to concern itself with what object formatter is being used.

Instead of the current test:

* write a test that verifies that the `assert` function returns `true` when we expect it, that the `assertMessage`, `refuteMessage `, `expectation` and `values` are set to expected values
* write a test that shows that `referee.js` is using `is-map` correctly
* and then finally write a test that verifies that `interpolateProperties` is using `util.inspect` as the object formatter

I'll give that a go, and push a commit that does this for only `isMap`. Then we can see what that looks like.

mroderick commented:
@mantoni I've pushed fresh commits that shows the new approach

* replaced `format` with one that uses `util.inspect`
* converted the test for `isMap` from a functional test to a unit test
* added a test to show that `isMap` is exported on both `referee.assert` and `referee.refute` 

There are still failing tests, don't mind those just now ... I can fix those later, if we agree on the approach.

What do you think?

mroderick commented:
Replaced by #183
